graphql/src/main/java/com/scalar/db/graphql/datafetcher/TransactionInstrumentation.java

```diff
@@ -0,0 +1,191 @@
+  private final DistributedTransactionManager transactionManager;
+
+  public TransactionInstrumentation(DistributedTransactionManager transactionManager) {
+    this.transactionManager = transactionManager;
+    this.activeTransactions =
+        new ActiveExpiringMap<>(
+            TRANSACTION_LIFETIME_MILLIS,
+            TRANSACTION_EXPIRATION_INTERVAL_MILLIS,
+            t -> {
+              LOGGER.warn("the transaction {} is expired", t.getId());
+              try {
+                t.abort();
+              } catch (AbortException e) {
+                LOGGER.warn("an error happened when aborting the transaction", e);
+              }
+            });
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginExecuteOperation(
+      InstrumentationExecuteOperationParameters parameters) {
+    ExecutionContext executionContext = parameters.getExecutionContext();
+    GraphQLContext graphQLContext = executionContext.getGraphQLContext();
+
+    List<Directive> transactionDirectives =
+        executionContext
+            .getOperationDefinition()
+            .getDirectives(Constants.TRANSACTION_DIRECTIVE_NAME);
+
+    if (transactionDirectives.isEmpty()) {
+      return SimpleInstrumentationContext.noOp();
+    }
+
+    Directive directive = transactionDirectives.get(0); // @transaction is not repeatable
+    Argument txIdArg = directive.getArgument(Constants.TRANSACTION_DIRECTIVE_TX_ID_ARGUMENT_NAME);
+    Argument commitArg =
+        directive.getArgument(Constants.TRANSACTION_DIRECTIVE_COMMIT_ARGUMENT_NAME);
+
+    DistributedTransaction transaction;
+    if (txIdArg == null) {
+      // start a new transaction
+      try {
+        transaction = transactionManager.start();
+      } catch (TransactionException e) {
+        LOGGER.warn("failed to start transaction", e);
+        throw new AbortExecutionException(e);
+      }
+    } else {
+      // continue an existing transaction
+      String txId = ((StringValue) txIdArg.getValue()).getValue();
+      transaction =
+          activeTransactions
+              .get(txId)
+              .orElseThrow(
+                  () -> {
+                    String message =
+                        String.format(
+                            "The specified transaction %s is not found."
+                                + " It might have been expired",
+                            txId);
+                    LOGGER.warn(message);
+                    return new AbortExecutionException(message);
+                  });
+    }
+    graphQLContext.put(Constants.CONTEXT_TRANSACTION_KEY, transaction);
+
+    boolean isCommitTrue =
+        commitArg != null
+            && commitArg.getValue() != null
+            && ((BooleanValue) commitArg.getValue()).isValue();
+    if (isCommitTrue) {
+      return SimpleInstrumentationContext.whenCompleted(
+          (executionResult, throwable) -> {
+            graphQLContext.delete(Constants.CONTEXT_TRANSACTION_KEY);
```

Comment:
For these two lines, it is OK to delete and remove here before committing or aborting?
(Is it OK even if it crashes after remove is called?)

Reply:
I think it is OK. The transaction object is removed in the beginning here because it is assumed that once the user sends a `commit` or `abort` request, the transaction should be finished and not be able to be continued in the next request.
In the normal case, either `commit` or `abort` is executed in this block. If the runtime crashes before committing or aborting, the transaction object is being not accessible, then it should eventually fail. (Is my understanding correct?)